storage-api/src/java/org/apache/hive/common/util/BloomKFilter.java

```diff
@@ -362,16 +379,178 @@ public static void mergeBloomFilterBytes(
+
+    for (int w = 0; w < workers.length; w++) {
+      int modifiedStart = START_OF_SERIALIZED_LONGS + w * elementPerBatch;
+      int modifiedLength = (w == workers.length - 1)
+        ? length - (START_OF_SERIALIZED_LONGS + w * elementPerBatch) : elementPerBatch;
+
+      ElementWrapper wrapper =
+          new ElementWrapper(bytes, start, length, modifiedStart, modifiedLength);
+      workers[w].add(wrapper);
+    }
+  }
+
+  public static byte[] getInitialBytes(long expectedEntries) {
+    ByteArrayOutputStream bytesOut = null;
+    try {
+      bytesOut = new ByteArrayOutputStream();
+      BloomKFilter bf = new BloomKFilter(expectedEntries);
+      BloomKFilter.serialize(bytesOut, bf);
+      return bytesOut.toByteArray();
+    } catch (Exception err) {
+      throw new IllegalArgumentException("Error creating aggregation buffer", err);
+    } finally {
+      IOUtils.closeStream(bytesOut);
+    }
+  }
+
+  public static class ElementWrapper {
+    public byte[] bytes;
+    public int start;
+    public int length;
+    public int modifiedStart;
+    public int modifiedLength;
+
+    public ElementWrapper(byte[] bytes, int start, int length, int modifiedStart, int modifiedLength) {
+      this.bytes = bytes;
+      this.start = start;
+      this.length = length;
+      this.modifiedStart = modifiedStart;
+      this.modifiedLength = modifiedLength;
+    }
+  }
+
+  private static class BloomFilterMergeWorker implements Runnable {
+    Queue<ElementWrapper> queue = new LinkedBlockingDeque<>();
+    private ExecutorService executor;
+
+    private byte[] bfAggregation;
+    private int bfAggregationStart;
+    private int bfAggregationLength;
+
+    public BloomFilterMergeWorker(ExecutorService executor, byte[] bfAggregation, int bfAggregationStart, int bfAggregationLength) {
+      this.executor = executor;
+      this.bfAggregation = bfAggregation;
+      this.bfAggregationStart = bfAggregationStart;
+      this.bfAggregationLength = bfAggregationLength;
+    }
+
+    public void add(ElementWrapper wrapper) {
+      queue.add(wrapper);
+    }
+
+    @Override
+    public void run() {
+      while (!executor.isTerminated() && !queue.isEmpty()) {
+        ElementWrapper currentBf = queue.poll();
+        if (currentBf != null) {
+          // at this point we have a currentBf wrapper which contains the whole byte[] of the
+          // serialized bloomfilter, but we only want to merge a modified "start -> start+length"
+          // part of it, which is pointed by modifiedStart/modifiedLength fields by ElementWrapper
+          BloomKFilter.mergeBloomFilterBytes(bfAggregation, bfAggregationStart, bfAggregationLength,
+              currentBf.bytes, currentBf.start, currentBf.length, currentBf.modifiedStart,
+              currentBf.modifiedStart + currentBf.modifiedLength);
+        } else {
+          try {
+            Thread.sleep(10); // relax this thread while the queue is empty
```

Comment:
Why is this needed? "splitVectorAcrossWorkers" seems to be adding all data before submitting to threadPool. IAC, it would be good to try with "queue::take()" (with additional change) which is a blocking call, to avoid "Thread.sleep".

Reply:
thanks @rbalamohan! that makes sense, I'll change and test on cluster